llpc/patch/llpcPatchIntrinsicSimplify.cpp

```diff
@@ -0,0 +1,416 @@
+
+    bool floatCoord = false;
+
+    for (uint32_t operandIndex : coordOperandIndices)
+    {
+        Value* const pCoord = pIntrinsicCall->getOperand(operandIndex);
+        // If the values are not derived from 16-bit values, we cannot optimize.
+        if (CanSafelyConvertTo16Bit(pCoord) == false)
+        {
+            return nullptr;
+        }
+
+        LLPC_ASSERT(operandIndex == coordOperandIndices[0]
+            || floatCoord == pCoord->getType()->isFloatingPointTy());
+        floatCoord = pCoord->getType()->isFloatingPointTy();
+    }
+
+    Type* const pCoordType = floatCoord ? Type::getHalfTy(*m_pContext) : Type::getInt16Ty(*m_pContext);
+
+    Function* const pIntrinsic = Intrinsic::getDeclaration(m_pFunction->getParent(),
+                                                           pIntrinsicCall->getIntrinsicID(),
+                                                           {pIntrinsicCall->getType(), pCoordType});
+
+    LLPC_ASSERT(pIntrinsic != nullptr);
+
+    SmallVector<Value*, 8> args(pIntrinsicCall->arg_operands());
+
+    IRBuilder<> builder(pIntrinsicCall);
+
+    for (uint32_t operandIndex : coordOperandIndices)
+    {
+        args[operandIndex] = ConvertTo16Bit(pIntrinsicCall->getOperand(operandIndex), builder);
+    }
+
+    return builder.CreateCall(pIntrinsic, args);
+}
+
+// =====================================================================================================================
+// Simplify a trigonometric intrinsic.
+Value* PatchIntrinsicSimplify::SimplifyTrigonometric(
+    IntrinsicInst* const pIntrinsicCall // [in] The intrinsic call to simplify
+    ) const
+{
+    // The sin and cos function in the hardware are dividing by 2*PI beforehand.
+    // This means      sin(x * 2 * PI) = amdgcn.sin(x)
+    // and                      sin(x) = amdgcn.sin(x / (2 * PI))
+    // We can switch to using our amdgcn trignonometric functions directly if the input conforms to the pattern:
+    // <trigonometric-function>(x * (2 * PI))
+    // <trigonometric-function>(x / (1 / (2 * PI)))
+
+    BinaryOperator* const pBinOp = dyn_cast<BinaryOperator>(pIntrinsicCall->getOperand(0));
+
+    // If the clamped value was not a binary operator, bail.
+    if (pBinOp == nullptr)
+    {
+        return nullptr;
+    }
+
+    ConstantFP* const pConstMultiplicator = dyn_cast<ConstantFP>(pBinOp->getOperand(1));
+
+    // If the multiplicator was not a constant, bail.
+    if (pConstMultiplicator == nullptr)
+    {
+        return nullptr;
+    }
+
+    APFloat multiplicator(pConstMultiplicator->getValueAPF());
+
+    bool losesInfo = false;
+
+    switch (pBinOp->getOpcode())
+    {
+    case BinaryOperator::FMul:
+        break;
+    case BinaryOperator::FDiv:
+    {
+        APFloat one(1.0);
+        one.convert(multiplicator.getSemantics(), APFloat::rmTowardZero, &losesInfo);
+        multiplicator = one / multiplicator;
+        break;
+    }
+    default:
+        return nullptr;
+    }
+
+    APFloat pi(M_PI);
```

Comment:
Can the value of `M_PI` be host-dependent? Can it change the codegen?

Reply:
Without this pass, we would emit code like `amdgcn.sin(a * (2*M_PI) / (2*M_PI))`, where one `M_PI` originates from the shader and the other one from LLVM.
This pass compares the `2*M_PI` from the shader with `2*M_PI` from llpc and removes it, resulting in `amdgcn.sin(a)`
The AMDGPU backend in LLVM uses the same constant as llpc, so even if `M_PI` is host-dependent, it should not make things worse.